File: functions/src/logger.ts
/**
 * Shared Logger Utility for Cloud Functions
 * 
 * Provides consistent logging across all Cloud Functions with automatic
 * Sentry integration for errors and warnings.
 */

import { Sentry } from './sentry';

/**
 * Logger interface for Cloud Functions
 * Maintains console output for Firebase Console logs while also sending
 * errors and warnings to Sentry for centralized monitoring
 */
export const logger = {
  /**
   * Log informational messages
   * Only sends to console, not to Sentry
   */
  info: (message: string, context: Record<string, unknown>) => {
    console.log(`[INFO] ${message}`, JSON.stringify(context));
  },

  /**
   * Log error messages
   * Sends to both console and Sentry with full context
   */
  error: (message: string, context: Record<string, unknown>) => {
    console.error(`[ERROR] ${message}`, JSON.stringify(context));
    
    // Send to Sentry with context
    Sentry.captureException(new Error(message), {
      level: 'error',
      extra: context,
    });
  },

  /**
   * Log warning messages
   * Sends to both console and Sentry with warning level
   */
  warn: (message: string, context: Record<string, unknown>) => {
    console.warn(`[WARN] ${message}`, JSON.stringify(context));
    
    // Send to Sentry as warning
    Sentry.captureMessage(message, {
      level: 'warning',
      extra: context,
    });
  },

  /**
   * Log debug messages
   * Only sends to console, not to Sentry
   */
  debug: (message: string, context: Record<string, unknown>) => {
    console.log(`[DEBUG] ${message}`, JSON.stringify(context));
  },
};

File: functions-python/logger.py
"""
Shared Logger Utility for Cloud Functions

Provides consistent logging across all Cloud Functions with automatic
Sentry integration for errors and warnings.
"""

import logging
from typing import Any

import sentry_sdk  # type: ignore

# Get logger instance
_logger = logging.getLogger(__name__)


def info(message: str, context: dict[str, Any]) -> None:
    """
    Log informational messages.
    Only sends to console, not to Sentry.
    """
    _logger.info(f"{message} | {context}")


def error(message: str, context: dict[str, Any]) -> None:
    """
    Log error messages.
    Sends to both console and Sentry with full context.
    """
    _logger.error(f"{message} | {context}")
    
    # Send to Sentry with context using scope
    with sentry_sdk.push_scope() as scope:  # type: ignore
        for key, value in context.items():
            scope.set_extra(key, value)  # type: ignore
        sentry_sdk.capture_exception(Exception(message))  # type: ignore


def warn(message: str, context: dict[str, Any]) -> None:
    """
    Log warning messages.
    Sends to both console and Sentry with warning level.
    """
    _logger.warning(f"{message} | {context}")
    
    # Send to Sentry as warning using scope
    with sentry_sdk.push_scope() as scope:  # type: ignore
        for key, value in context.items():
            scope.set_extra(key, value)  # type: ignore
        sentry_sdk.capture_message(message, level="warning")  # type: ignore


def debug(message: str, context: dict[str, Any]) -> None:
    """
    Log debug messages.
    Only sends to console, not to Sentry.
    """
    _logger.debug(f"{message} | {context}")

